File: 06-Compiler/02-Scanner/scanner.cpp
#include <cctype>
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <string>
#include <vector>
#ifdef _WIN32
#include <windows.h>
#endif

using namespace std;

struct Token
{
    string type;
    string lexeme;
    int line;
    int column;
};

static const map<string, string> KEYWORDS = {
    {"int", "INT"},
    {"float", "FLOAT"},
    {"void", "VOID"},
    {"if", "IF"},
    {"else", "ELSE"},
    {"while", "WHILE"},
    {"return", "RETURN"},
    {"input", "INPUT"},
    {"print", "PRINT"},
};

static bool isLetter(char ch)
{
    return isalpha(static_cast<unsigned char>(ch)) != 0;
}

static bool isDigit(char ch)
{
    return isdigit(static_cast<unsigned char>(ch)) != 0;
}

static bool isValueEnd(const string &type)
{
    return type == "ID" || type == "NUM" || type == "FLO" || type == "RPA" || type == "RBK";
}

static bool canReadSignedNumber(char sign, const string &source, size_t pos, const vector<Token> &tokens)
{
    if ((sign != '+' && sign != '-') || pos + 1 >= source.size())
    {
        return false;
    }

    if (!isDigit(source[pos + 1]) && source[pos + 1] != '.')
    {
        return false;
    }

    return tokens.empty() || !isValueEnd(tokens.back().type);
}

static string aliasForSampleOutput(const string &type)
{
    if (type == "LPA")
        return "LPAR";
    if (type == "RPA")
        return "RPAR";
    if (type == "SCO")
        return "SEMI";
    if (type == "FLO")
        return "FLOAT";
    return type;
}

static void addToken(vector<Token> &tokens, const string &type, const string &lexeme, int line, int column)
{
    tokens.push_back({type, lexeme, line, column});
}

static Token readIdentifierOrKeyword(const string &source, size_t &pos, int line, int column)
{
    size_t start = pos++;
    while (pos < source.size() && (isLetter(source[pos]) || isDigit(source[pos])))
    {
        ++pos;
    }

    string lexeme = source.substr(start, pos - start);
    auto keyword = KEYWORDS.find(lexeme);
    return {keyword == KEYWORDS.end() ? "ID" : keyword->second, lexeme, line, column};
}

static Token readNumber(const string &source, size_t &pos, int line, int column)
{
    size_t start = pos;

    if (source[pos] == '+' || source[pos] == '-')
    {
        ++pos;
    }

    bool hasDot = false;
    bool hasExp = false;

    while (pos < source.size() && isDigit(source[pos]))
    {
        ++pos;
    }

    if (pos < source.size() && source[pos] == '.')
    {
        hasDot = true;
        ++pos;
        while (pos < source.size() && isDigit(source[pos]))
        {
            ++pos;
        }
    }

    if (pos < source.size() && (source[pos] == 'e' || source[pos] == 'E'))
    {
        size_t expStart = pos;
        ++pos;

        if (pos < source.size() && (source[pos] == '+' || source[pos] == '-'))
        {
            ++pos;
        }

        bool hasExpDigit = false;
        while (pos < source.size() && isDigit(source[pos]))
        {
            hasExpDigit = true;
            ++pos;
        }

        if (hasExpDigit)
        {
            hasExp = true;
        }
        else
        {
            pos = expStart;
        }
    }

    string lexeme = source.substr(start, pos - start);
    return {(hasDot || hasExp) ? "FLO" : "NUM", lexeme, line, column};
}

static vector<Token> scan(const string &source)
{
    vector<Token> tokens;
    size_t pos = 0;
    int line = 1;
    int column = 1;

    while (pos < source.size())
    {
        char ch = source[pos];

        if (ch == '\r')
        {
            ++pos;
            continue;
        }
        if (ch == '\n')
        {
            ++line;
            column = 1;
            ++pos;
            continue;
        }
        if (isspace(static_cast<unsigned char>(ch)))
        {
            ++column;
            ++pos;
            continue;
        }

        if (ch == '/' && pos + 1 < source.size() && source[pos + 1] == '/')
        {
            while (pos < source.size() && source[pos] != '\n')
            {
                ++pos;
                ++column;
            }
            continue;
        }

        int tokenLine = line;
        int tokenColumn = column;
        size_t oldPos = pos;

        if (isLetter(ch))
        {
            tokens.push_back(readIdentifierOrKeyword(source, pos, tokenLine, tokenColumn));
        }
        else if (isDigit(ch) || (ch == '.' && pos + 1 < source.size() && isDigit(source[pos + 1])) ||
                 canReadSignedNumber(ch, source, pos, tokens))
        {
            tokens.push_back(readNumber(source, pos, tokenLine, tokenColumn));
        }
        else
        {
            if (pos + 1 < source.size())
            {
                string two = source.substr(pos, 2);
                if (two == "+=")
                {
                    addToken(tokens, "AAS", two, tokenLine, tokenColumn);
                    pos += 2;
                }
                else if (two == "++")
                {
                    addToken(tokens, "AAA", two, tokenLine, tokenColumn);
                    pos += 2;
                }
                else if (two == "<=" || two == "==" || two == ">=" || two == "!=")
                {
                    addToken(tokens, "ROP", two, tokenLine, tokenColumn);
                    pos += 2;
                }
                else if (two == "&&" || two == "||")
                {
                    addToken(tokens, "BOP", two, tokenLine, tokenColumn);
                    pos += 2;
                }
            }

            if (oldPos == pos)
            {
                switch (ch)
                {
                case '+':
                    addToken(tokens, "ADD", "+", tokenLine, tokenColumn);
                    break;
                case '-':
                    addToken(tokens, "SUB", "-", tokenLine, tokenColumn);
                    break;
                case '*':
                    addToken(tokens, "MUL", "*", tokenLine, tokenColumn);
                    break;
                case '/':
                    addToken(tokens, "DIV", "/", tokenLine, tokenColumn);
                    break;
                case '<':
                case '>':
                    addToken(tokens, "ROP", string(1, ch), tokenLine, tokenColumn);
                    break;
                case '!':
                    addToken(tokens, "BOP", "!", tokenLine, tokenColumn);
                    break;
                case '=':
                    addToken(tokens, "ASG", "=", tokenLine, tokenColumn);
                    break;
                case '(':
                    addToken(tokens, "LPA", "(", tokenLine, tokenColumn);
                    break;
                case ')':
                    addToken(tokens, "RPA", ")", tokenLine, tokenColumn);
                    break;
                case '[':
                    addToken(tokens, "LBK", "[", tokenLine, tokenColumn);
                    break;
                case ']':
                    addToken(tokens, "RBK", "]", tokenLine, tokenColumn);
                    break;
                case '{':
                    addToken(tokens, "LBR", "{", tokenLine, tokenColumn);
                    break;
                case '}':
                    addToken(tokens, "RBR", "}", tokenLine, tokenColumn);
                    break;
                case ',':
                    addToken(tokens, "CMA", ",", tokenLine, tokenColumn);
                    break;
                case ':':
                    addToken(tokens, "COL", ":", tokenLine, tokenColumn);
                    break;
                case ';':
                    addToken(tokens, "SCO", ";", tokenLine, tokenColumn);
                    break;
                default:
                    addToken(tokens, "ERROR", string(1, ch), tokenLine, tokenColumn);
                    break;
                }
                ++pos;
            }
        }

        column += static_cast<int>(pos - oldPos);
    }

    return tokens;
}

static string stripBom(string text);

static string readAll(const string &filename)
{
    ifstream fin(filename.c_str(), ios::in | ios::binary);
    if (!fin.is_open())
    {
        return "";
    }

    stringstream buffer;
    buffer << fin.rdbuf();
    string text = buffer.str();
    return stripBom(text);
}

static string stripBom(string text)
{
    if (text.size() >= 3 &&
        static_cast<unsigned char>(text[0]) == 0xEF &&
        static_cast<unsigned char>(text[1]) == 0xBB &&
        static_cast<unsigned char>(text[2]) == 0xBF)
    {
        text.erase(0, 3);
    }
    return text;
}

static void printTypesOnly(const vector<Token> &tokens)
{
    for (const Token &token : tokens)
    {
        cout << aliasForSampleOutput(token.type) << '\n';
    }
}

static void printPairs(const vector<Token> &tokens)
{
    for (const Token &token : tokens)
    {
        cout << "(" << token.type << ", " << token.lexeme << ")\n";
    }
}

static void showMenu()
{
    cout << "===== Scanner 词法分析系统 =====\n";
    cout << "1. 分析符号串类型\n";
    cout << "2. 输出程序语句的token流\n";
    cout << "3. 解析代码文件的token流\n";
    cout << "注：将终端输出到文件需使用：scanner.exe 源文件 [输出文件]\n";
    cout << "请输入你的选择: ";
}

static void runMode1()
{
    int n;
    cout << "请输入符号串个数n: ";
    cin >> n;
    cout << "请输入" << n << "个用空格分隔的符号串: ";
    for (int i = 0; i < n; ++i)
    {
        string word;
        cin >> word;
        vector<Token> tokens = scan(word);
        if (tokens.size() == 1 && tokens[0].lexeme == word)
        {
            cout << aliasForSampleOutput(tokens[0].type) << '\n';
        }
        else
        {
            cout << "ERROR\n";
        }
    }
}

static void runMode2()
{
    string line;
    getline(cin, line);
    cout << "请输入一行源程序语句: ";
    if (line.empty())
    {
        getline(cin, line);
    }
    printTypesOnly(scan(line));
}

static void runMode3()
{
    string filename;
    cout << "请输入源程序文件名: ";
    cin >> filename;
    string source = readAll(filename);
    if (source.empty())
    {
        cerr << "Cannot read file: " << filename << '\n';
        return;
    }
    printPairs(scan(source));
}

int main(int argc, char *argv[])
{
#ifdef _WIN32
    SetConsoleOutputCP(CP_UTF8);
    SetConsoleCP(CP_UTF8);
#endif

    if (argc >= 2)
    {
        string source = readAll(argv[1]);
        if (source.empty())
        {
            cerr << "Cannot read file: " << argv[1] << '\n';
            return 1;
        }

        if (argc >= 3)
        {
            ofstream fout(argv[2]);
            streambuf *old = cout.rdbuf(fout.rdbuf());
            printPairs(scan(source));
            cout.rdbuf(old);
        }
        else
        {
            printPairs(scan(source));
        }
        return 0;
    }

    showMenu();
    string modeText;
    cin >> modeText;
    modeText = stripBom(modeText);
    int mode = 0;
    stringstream modeStream(modeText);
    modeStream >> mode;
    if (mode == 1)
    {
        runMode1();
    }
    else if (mode == 2)
    {
        runMode2();
    }
    else if (mode == 3)
    {
        runMode3();
    }
    else
    {
        cerr << "Mode must be 1, 2 or 3.\n";
        return 1;
    }

    return 0;
}

File: 06-Compiler/02-Scanner/ComplicationWeb/scanner.js
const scannerSample = `void qsort(int low, int high)
{
    void swap(int i,j) {void qsort(int low, int high)}
    int k;
    int parti(int lowhigh)
    {
    void swap(int i,j){
    int x;x=a[i];a[i]=a[j];a[i]=x}//end of swap
    }
}`;

const keywords = new Map([
  ["int", "INT"],
  ["float", "FLOAT"],
  ["void", "VOID"],
  ["if", "IF"],
  ["else", "ELSE"],
  ["while", "WHILE"],
  ["return", "RETURN"],
  ["input", "INPUT"],
  ["print", "PRINT"]
]);

function isLetter(ch) {
  return /^[A-Za-z]$/.test(ch);
}

function isDigit(ch) {
  return /^[0-9]$/.test(ch);
}

function isValueEnd(type) {
  return ["ID", "NUM", "FLO", "RPA", "RBK"].includes(type);
}

function canReadSignedNumber(source, pos, tokens) {
  const ch = source[pos];
  if (ch !== "+" && ch !== "-") return false;
  const next = source[pos + 1] || "";
  if (!isDigit(next) && next !== ".") return false;
  return tokens.length === 0 || !isValueEnd(tokens[tokens.length - 1].type);
}

function scanSource(source) {
  const tokens = [];
  let pos = 0;

  function push(type, lexeme) {
    tokens.push({ type, lexeme });
  }

  while (pos < source.length) {
    const ch = source[pos];

    if (/\s/.test(ch)) {
      pos++;
      continue;
    }

    if (ch === "/" && source[pos + 1] === "/") {
      while (pos < source.length && source[pos] !== "\n") pos++;
      continue;
    }

    if (isLetter(ch)) {
      const start = pos++;
      while (pos < source.length && (isLetter(source[pos]) || isDigit(source[pos]))) pos++;
      const lexeme = source.slice(start, pos);
      push(keywords.get(lexeme) || "ID", lexeme);
      continue;
    }

    if (isDigit(ch) || (ch === "." && isDigit(source[pos + 1] || "")) || canReadSignedNumber(source, pos, tokens)) {
      const start = pos;
      if (source[pos] === "+" || source[pos] === "-") pos++;
      while (isDigit(source[pos] || "")) pos++;

      let hasDot = false;
      let hasExp = false;
      if (source[pos] === ".") {
        hasDot = true;
        pos++;
        while (isDigit(source[pos] || "")) pos++;
      }

      if (source[pos] === "e" || source[pos] === "E") {
        const expStart = pos++;
        if (source[pos] === "+" || source[pos] === "-") pos++;
        let hasExpDigit = false;
        while (isDigit(source[pos] || "")) {
          hasExpDigit = true;
          pos++;
        }
        if (hasExpDigit) hasExp = true;
        else pos = expStart;
      }

      const lexeme = source.slice(start, pos);
      push(hasDot || hasExp ? "FLO" : "NUM", lexeme);
      continue;
    }

    const two = source.slice(pos, pos + 2);
    if (two === "+=") {
      push("AAS", two);
      pos += 2;
      continue;
    }
    if (two === "++") {
      push("AAA", two);
      pos += 2;
      continue;
    }
    if (["<=", "==", ">=", "!="].includes(two)) {
      push("ROP", two);
      pos += 2;
      continue;
    }
    if (two === "&&" || two === "||") {
      push("BOP", two);
      pos += 2;
      continue;
    }

    const single = {
      "+": "ADD",
      "-": "SUB",
      "*": "MUL",
      "/": "DIV",
      "<": "ROP",
      ">": "ROP",
      "!": "BOP",
      "=": "ASG",
      "(": "LPA",
      ")": "RPA",
      "[": "LBK",
      "]": "RBK",
      "{": "LBR",
      "}": "RBR",
      ",": "CMA",
      ":": "COL",
      ";": "SCO"
    };

    push(single[ch] || "ERROR", ch);
    pos++;
  }

  return tokens;
}

function typeAlias(type) {
  if (type === "LPA") return "LPAR";
  if (type === "RPA") return "RPAR";
  if (type === "SCO") return "SEMI";
  if (type === "FLO") return "FLOAT";
  return type;
}

document.getElementById("loadScannerSampleBtn").addEventListener("click", () => {
  document.getElementById("sourceInput").value = scannerSample;
  showToast("测试程序已载入", "success");
});

document.getElementById("scanPairsBtn").addEventListener("click", () => {
  const tokens = scanSource(document.getElementById("sourceInput").value);
  document.getElementById("scannerOutput").textContent = tokens
    .map(token => `(${token.type}, ${token.lexeme})`)
    .join("\n");
});

document.getElementById("scanTypesBtn").addEventListener("click", () => {
  const tokens = scanSource(document.getElementById("sourceInput").value);
  document.getElementById("scannerOutput").textContent = tokens.map(token => typeAlias(token.type)).join("\n");
});

window.addEventListener("DOMContentLoaded", () => {
  document.getElementById("sourceInput").value = scannerSample;
});
